honeybadgermpc/progs/mimc.py

```diff
@@ -30,9 +38,36 @@
     return inp + k
 
 
-def mimc_plain(x, k):
-    inp = x
+async def mimc_mpc_batch(context, xs, k):
+    """
+    MiMC block cipher encryption encrypts blocks of message xs with secret key k,
+    where xs are a list of shared secrets, k is an element of F_p
+    """
+    pp_elements = PreProcessedElements()
+
+    # def cubing_share_array(): [x1,..., xK] -> [x1^3,..., xK^3]
+    async def cubing_share_array(xs):
+        rs = [[], [], []]
+        ys_, x3s = [], []
+
+        for x in xs:
+            r1, r2, r3 = pp_elements.get_cube(context)
+            y = await (x - r1).open()
```

Comment:
I think we can do this better with a gather--
```
rs = rs_sq = cs_cube = []
for _ in range(len(xs)):
    r1, r2, r3 =pp_elements.get_cube(context)
    rs.append(r1)
    rs_sq.append(r2)
    rs_cube.append(r3)
rs = context.Share_Array(rs)
rs_sq = context.Share_Array(rs_sq)
rs_cub = context.Share_Array(rs_cub)
ys = await (await(xs - rs).open())
etc....
```

Reply:
Good idea.